composer/datasets/synthetic_lm.py

```diff
@@ -0,0 +1,231 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+from __future__ import annotations
+
+import json
+import random
+import string
+import textwrap
+from os.path import join
+from tempfile import mkdtemp
+from typing import TYPE_CHECKING, NamedTuple, Optional
+
+if TYPE_CHECKING:
+    import tokenizers.decoders as decoders
+    import tokenizers.models as tokenizers_models
+    import tokenizers.normalizers as normalizers
+    import tokenizers.pre_tokenizers as pre_tokenizers
+    from datasets import Dataset
+    from transformers import PreTrainedTokenizer
+
+
+class SyntheticTokenizerParams(NamedTuple):
+    tokenizer_model: tokenizers_models.Model
+    normalizer: normalizers.Normalizer
+    pre_tokenizer: pre_tokenizers.PreTokenizer
+    decoder: decoders.Decoder
+    initial_alphabet: list
+    special_tokens: list
+    pad_token: str
+    trainer_cls: type
+    tokenizer_cls: type
+
+
+def _generate_bert_tokenizer_params(dataset) -> SyntheticTokenizerParams:
+    try:
+        import tokenizers.decoders as decoders
```

Comment:
this is pretty verbose and repeated multiple times in this file. i think we need to come up with a general pattern for files that need to be imported, but have these optional dependencies. cc: @ravi-mosaicml or @dblalock for suggestions?

Reply:
Strong agree here, I'm pretty interested in fixing this!

Reply:
I don't know what the best way of dealing with this is. I could imagine a few solutions:

1.  We have a `composer.tokenizers` that lazily tries to import hf tokenizers in its getattr.
2.  We have a singleton imported as `composer.tokenizers` that does the same thing.
3.  We have a `composer.tokenizers` that tries to import hf tokenizers, but instead creates a Mock if that fails. The Mock throws the informative ImportError below whenever it's called.

3 seems the least bad, but none of these seem ideal.

Reply:
Discussed offline: out of scope for this PR